library/psa_crypto.c

```diff
@@ -713,6 +713,29 @@ static psa_status_t psa_get_empty_key_slot( psa_key_handle_t handle,
     return( status );
 }
 
+/** Test whether a policy permits an algorithm.
+ *
+ * The caller must test usage flags separately.
+ */
+static int psa_key_policy_permits( const psa_key_policy_t *policy,
+                                   psa_algorithm_t alg )
+{
+    /* Common case: the policy only allows alg. */
+    if( alg == policy->alg )
+        return( 1 );
+    /* If policy->alg is a hash-and-sign with a wildcard for the hash,
+     * and alg is the same hash-and-sign family with any hash,
+     * then alg is compliant with policy->alg. */
+    if( PSA_ALG_IS_HASH_AND_SIGN( alg ) &&
+        PSA_ALG_SIGN_GET_HASH( policy->alg ) == PSA_ALG_ANY_HASH )
```

Comment:
This is slightly duplicated because `PSA_ALG_SIGN_GET_HASH` already includes `PSA_ALG_IS_HASH_AND_SIGN`.

Reply:
Could we remove `PSA_ALG_IS_HASH_AND_SIGN( alg ) &&` from this and have the same behavior?

Reply:
Operationally, given the current implementation, we could. But semantically, `PSA_ALG_SIGN_GET_HASH` has unspecified behavior if the argument is not a hash-and-sign algorithm. Let the compiler optimize.